samples/config_helper_vulkan.cc

```diff
@@ -764,26 +659,46 @@ void ConfigHelperVulkan::CreateVulkanDevice(
       static_cast<uint32_t>(required_extensions_in_char.size());
   info.ppEnabledExtensionNames = required_extensions_in_char.data();
 
-  assert(vkCreateDevice(vulkan_physical_device_, &info, nullptr,
-                        &vulkan_device_) == VK_SUCCESS);
+  if (vkCreateDevice(vulkan_physical_device_, &info, nullptr,
+                     &vulkan_device_) != VK_SUCCESS) {
+    return amber::Result("Unable to create vulkan device");
+  }
+  return {};
 }
 
-std::unique_ptr<amber::EngineConfig> ConfigHelperVulkan::CreateConfig(
+amber::Result ConfigHelperVulkan::CreateConfig(
     const std::vector<std::string>& required_features,
-    const std::vector<std::string>& required_extensions) {
-  auto required_vulkan_features = NamesToVulkanFeatures(required_features);
+    const std::vector<std::string>& required_extensions,
+    std::unique_ptr<amber::EngineConfig>* cfg_holder) {
+  VkPhysicalDeviceFeatures required_vulkan_features = {};
+  amber::Result r =
+      NamesToVulkanFeatures(required_features, &required_vulkan_features);
+  if (!r.IsSuccess())
+    return r;
+
+  r = CreateVulkanInstance();
+  if (!r.IsSuccess())
+    return r;
+
+  r = CreateDebugReportCallback();
+  if (!r.IsSuccess())
+    return r;
+
+  r = ChooseVulkanPhysicalDevice(required_vulkan_features, required_extensions);
+  if (!r.IsSuccess())
+    return r;
+
+  r = CreateVulkanDevice(required_vulkan_features, required_extensions);
+  if (!r.IsSuccess())
+    return r;
 
-  CreateVulkanInstance();
-  CreateDebugReportCallback();
-  ChooseVulkanPhysicalDevice(required_vulkan_features, required_extensions);
-  CreateVulkanDevice(required_vulkan_features, required_extensions);
   vkGetDeviceQueue(vulkan_device_, vulkan_queue_family_index_, 0,
                    &vulkan_queue_);
 
-  std::unique_ptr<amber::EngineConfig> cfg_holder =
+  *cfg_holder =
```

Comment:
I would have gone for cfg_holder->reset( the new'd pointer ).

But this is fine.